designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
```

Comment:
What are your thoughts about mounting a bucket (or a user-specified key prefix within a bucket) [to the filesystem](https://code.visualstudio.com/api/references/vscode-api#FileSystemProvider)? This would provide an idiomatic way to explore object based data, where most of the user operations are already known to be file-esque. This potentially simplifies the object explorer wrt S3 too.

Reply:
Definitely interesting. Filesystems are hard™ so this may be a big project.

Reply:
I agree with the sentiment. I'd be interested in hearing about a prototype on this. Perhaps VS Code takes care of enough of the heavy lifting for extension developers.

Reply:
What's the justification for the departure on the JetBrains / Visual Studio pattern of "opening the bucket" in a separate editor window and browsing from there? I think integrating the whole experience into the top-level explorer is going to be clunky - especially for large buckets.

I see the AWS Explorer as an "entry-point" where you find some resource you want to interact with (in this case the Bucket is the resource) once you've located it, further interactions happen in a separate visual space. What if I want to interact with two buckets at once? Scrolling up and down large buckets in the explorer is not an effective way to do that.

> Filesystems are hard™ so this may be a big project.

Agree - we tried this already with JetBrains and abandoned it.